test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java

```diff
@@ -0,0 +1,371 @@
+            log.debug("Create stream status {}", createStreamStatus);
+        }
+    }
+
+    void createWriters(ClientFactory clientFactory, final int writers, String scope, String stream) {
+        log.info("Client factory details {}", clientFactory.toString());
+        log.info("Creating {} writers", writers);
+        writerList = new ArrayList<>(writers);
+        log.info("Writers writing in the scope {}", scope);
+        CompletableFuture.runAsync(() -> {
+                    for (int i = 0; i < writers; i++) {
+                        log.info("Starting writer{}", i);
+                        final EventStreamWriter<Long> tmpWriter = clientFactory.createEventWriter(stream,
+                                new JavaSerializer<Long>(),
+                                EventWriterConfig.builder().retryAttempts(10).build());
+                        writerList.add(tmpWriter);
+                        writerFutureList.add(startWriting(tmpWriter));
+                    }
+                });
+    }
+
+    void createReaders(ClientFactory clientFactory, String readerGroupName, String scope,
+                                 ReaderGroupManager readerGroupManager, String stream, final int readers) {
+        log.info("Creating Reader group: {}, with readergroup manager using scope: {}", readerGroupName, scope);
+        readerGroupManager.createReaderGroup(readerGroupName, ReaderGroupConfig.builder().startingTime(0).build(),
+                Collections.singleton(stream));
+        log.info("Reader group name: {}, Reader group scope: {}, Online readers: {}",
+                readerGroupManager.getReaderGroup(readerGroupName).getGroupName(), readerGroupManager
+                        .getReaderGroup(readerGroupName).getScope(), readerGroupManager
+                        .getReaderGroup(readerGroupName).getOnlineReaders());
+        log.info("Creating {} readers", readers);
+        readerList = new ArrayList<>(readers);
+        log.info("Scope that is seen by readers {}", scope);
+
+        CompletableFuture.runAsync(() -> {
+            for (int i = 0; i < readers; i++) {
+                log.info("Starting reader: {}, with id: {}", i, readerName + i);
+                final EventStreamReader<Long> reader = clientFactory.createReader(readerName + i,
+                        readerGroupName,
+                        new JavaSerializer<Long>(),
+                        ReaderConfig.builder().build());
+                readerList.add(reader);
+                readerFutureList.add(startReading(reader));
+            }
+        });
+    }
+
+    void stopReadersAndWriters(ReaderGroupManager readerGroupManager, String readerGroupName) throws InterruptedException, ExecutionException {
+
+        log.info("Stop write flag status {}", testState.stopWriteFlag);
+        testState.stopWriteFlag.set(true);
+
+        log.info("Wait for writers execution to complete");
+        FutureHelpers.allOf(writerFutureList).get();
```

Comment:
Currently this is potentially non-thread safe depending on the actions of the subclassing tests.
This could be eliminated by creating a future in TestState (which is final) and then when the readers are generated you could call `FutureHelpers.completeAfter(()->FutureHelpers.allOf(writerFutures), testState.writersComplete);`
Then here it would only need to do `testState.writersComplete.get()`